core/solver/batch_solver.ipp

```diff
@@ -0,0 +1,132 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_SOLVER_BATCH_SOLVER_IPP_
+#define GKO_CORE_SOLVER_BATCH_SOLVER_IPP_
+
+
+#include <ginkgo/core/log/batch_convergence.hpp>
+#include <ginkgo/core/solver/batch_solver.hpp>
+
+
+#include "core/log/batch_logging.hpp"
+#include "core/matrix/batch_csr_kernels.hpp"
+#include "core/matrix/batch_dense_kernels.hpp"
+
+
+namespace gko {
+namespace solver {
+namespace batch {
+
+
+GKO_REGISTER_OPERATION(pre_diag_scale_system, batch_csr::pre_diag_scale_system);
+GKO_REGISTER_OPERATION(vec_scale, batch_dense::batch_scale);
+
+
+}
+
+
+struct BatchInfo {
```

Comment:
The problem is that the `logdata` object depends on the value type. The class definition of `EnableBatchSolver` cannot depend on the ValueType, because it's queried from the CRTP parameter `ConcreteSolver`. If the definition of `EnableBatchSolver` depends on the `ConcreteSolver`'s `value_type`, we get an incomplete definition.

Reply:
I see. Maybe we can keep `EnableBatchSolver` un-templated on `ValueType` but infer the value type from the `ConcreteSolver` for the `logdata` object with something like `ConcreteSolver::value_type` and just forward declare those, if not available ?

Reply:
The root issue here is
- We want to subclass `EnableBatchSolver` from `EnableBatchLinOp` (going with the mixin philosophy of a linear inheritance chain rather than a tree), so it needs the Concrete type.
- We cannot use `ConcreteType::value_type` in the interface definition of the class because `ConcreteType` itself depends on `EnableBatchSolver`.

I'm not sure what kind of forward declaration will help here. But I agree that this is a hack (void pointer) and it would be nice to have a better solution.

My earlier solution was to have an additional `ValueType` template parameter on `EnableBatchSolver`. That solves this problem, but then there is are two redundant value types - one coming from the direct template parameter and one coming from `ConcreteType::value_type`.

Reply:
You could have a base-class of `BatchLogData` that is independent of `value_type`, and classes inheriting from it, using the actual type (which we do with the Ginkgo logger). You could move the creation of the logger to each solver.

Inside the actual logger implementation, you only have to use `as<BatchDense<value_type>>` to convert it to the correct type (and throw an error if it mismatches). That way, you don't need to know which `value_type` you are using and you don't need a `void*` (you also would not need an additional layer to hide the `void*` since you could use the `gko::log::BatchLogger` or whatever you want to call it).